### PR TITLE
Split options from flags for annotate command

### PR DESCRIPTION
**Pull request by iamNoah1**
#### What type of PR is this?

This PR aims to split flags from options for the annotate command

#### What this PR does / why we need it:

see 
https://github.com/kubernetes/kubectl/issues/1276 

and 

https://github.com/kubernetes/kubectl/issues/1046

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:245]: We are using `flags.RestClientGetter` in here but we are also passing same object `cmdutil.Factory` as parameter?

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:245]: Why we are exporting `options.Namespace`?

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:161]: After checking annotate command, it relies on openapigetter and unstructuredClientForMapping which is defined in cmdutil.Factory. That's why, we can use cmdutil.Factory and remove `restclientgetter` at least for now.

**Review by ardaguclu — COMMENTED** [review]
Thanks for this PR @iamNoah1

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (iamNoah1) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:245]: I didn't know better :), Fixed both.

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:282]: It is also a good time to rename that function to `Run`

Comment (iamNoah1) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:282]: My first feeling was the same, somehow I don't like it. But is it really that bad? 

Things like `Local`, `Selector`, `FieldSelector`, `FileNameOptions` and some more are not available in Options. I mean, I could move them to the Options Struct again, but they are flags aren't they? 

I could also have them in both options and flags struct, like for `allnamespace`. I cannot say what is better. I am probably overlooking something.

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:170]: I think, all the flags might have a corresponding object in flags and this should not be in here. Because it looks to me that this is breaking the loose coupling and the simplicity of flow(get flags, get options from flags and options.run without flags)

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:283]: > Things like Local, Selector, FieldSelector, FileNameOptions and some more are not available in Options. I mean, I could move them to the Options Struct again, but they are flags aren't they?

Yes, they are definitely flag. But why we are not storing both objects in also flags and options and map them during `toOptions`?. I think there is no problem with having same fields in flag and option. As well as we have a clear simple command flow.

I understand your concern about managing same fields in different places. But if we pass `flags` into run function, in the future, PR owners will not bother themselves adding something into options, because they can just add flag and use it in `Run` function and inevitably options will become redundant.

@soltysh WDYT?

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (iamNoah1) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go]: @ardaguclu, of course, what you are saying makes totally sense. I tried to update accordingly :)

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:54]: What is the reason of these two DryRun fields(strategy and verifier) are in AnnotateFlags?

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:258]: Would it be better or possible if we move the rest of this complete function into `validate`?. Because as far as I can see, all the rest is validation.

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (iamNoah1) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:54]: That is very good question. I didn't know better. Do you think it makes more sense to leave them in AnnotateOptions? @ardaguclu

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (iamNoah1) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:258]: yeah this has been a tricky part. I decided to go this way, as I was looking at other PRs that split flags from options. If you are ok, and the others folks do not object I would to leave it as it is.

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate_test.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate_test.go:571]: forgot to remove this? :)

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:54]: I think, this(and DryRunVerifier) can be only in options.

**Review by ardaguclu — COMMENTED** [review]
Generally lgtm and left a few comments. I'll defer to approvers for a second eye. Thanks @iamNoah1

**Comment by ardaguclu** [comment]
Would you mind adding unit tests cases especially covering the validations like in here https://github.com/kubernetes/kubernetes/pull/112817/files#diff-9ccf80373137694902ef1403ddc2640d3511785056ff8ee07292611b95bdcdffR268-R286?

File: staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go:54]: Yeah, try to stick with the rule, if it's not a flag then most likely it should go to `*Options` struct.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
